src/util/builtinOperators.ts

```diff
@@ -336,6 +337,34 @@ function generateGFunction(rm: rbac.RoleManager): any {
   };
 }
 
+// generateSyncedGFunction is the synchronous factory method of the g(_, _) function.
+function generateSyncedGFunction(rm: DefaultRoleManager): any {
```

Comment:
This should depend on the interface, not implementation.

Reply:
We can new a new interface that extends RoleManager, then use that interface here.

Reply:
Currently I am not even sure why most RoleManager's functions are async. They do not `await` anything so even if some functions inside them were promises, they would not work anyway.

I guess if somebody used `.then` with `hasLink` it would be a breaking change... However this makes source code more confusing than it should be.

What should we do? I guess a new RoleManager class and interface could be created which is not async.

Reply:
> Currently I am not even sure why most RoleManager's functions are async. They do not await anything so even if some functions inside them were promises, they would not work anyway.

Some users store the role link on the database/file, so we need to use the async. 
Issue: https://github.com/casbin/node-casbin/issues/46

> What should we do? I guess a new RoleManager class and interface could be created which is not async.

Can we create a RoleManager that only provides the synced method?

Reply:
Hmmph. I think interface itself should cover all the possible solutions and should not force interface implementation to be async.

I kinda prefer this implementation referenced in the solution.
`getRoles(name: string, ...domain: string[]): Promise<string[]> | string[]`

I think one problem at the moment is that `syncedEnforcer` exists. It might be really confusing for a new user to find out that `syncedEnforcer.enforce()`,  `syncedEnforcer.enforceSync()`, `enforcer.enforceSync()` are all completely different things.

Overall the performance is the problem. I am not sure which way I would tackle this.

It would be fine to use just async function if it performed at acceptable level. I am puzzled by this.

Could the `privateEnforce`-function be fixed with something like [fastest promise wins](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any)? It feels so weird that we have to have separate functions for these.

Reply:
@nodece Added `syncedHasLink` as an optional parameter for `RoleManager` and changed all the functions to use the interface instead.

In my initial testing with database of 35 thousand rows, this would reduce the random read time to one third of what it is now.

Reply:
> I think one problem at the moment is that syncedEnforcer exists. It might be really confusing for a new user to find out that syncedEnforcer.enforce(), syncedEnforcer.enforceSync(), enforcer.enforceSync() are all completely different things.

`syncedEnforcer` is a thread-safe enforcer, `syncedEnforcer.enforceSync()` is a sync method.

> Could the privateEnforce-function be fixed with something like [fastest promise wins](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any)? It feels so weird that we have to have separate functions for these.

`privateEnforce`-function doesn't expose, which is called by all enforce.